include/dlaf/communication/communicator_pipeline.h

```diff
@@ -0,0 +1,141 @@
+#include <dlaf/communication/communicator_type.h>
+#include <dlaf/communication/index.h>
+
+namespace dlaf::comm {
+namespace internal {
+constexpr const dlaf::common::Ordering FULL_COMMUNICATOR_ORDER{dlaf::common::Ordering::RowMajor};
+}
+
+using CommunicatorPipelineReadOnlyWrapper = typename common::Pipeline<Communicator>::ReadOnlyWrapper;
+using CommunicatorPipelineReadWriteWrapper = typename common::Pipeline<Communicator>::ReadWriteWrapper;
+using CommunicatorPipelineReadOnlySender = typename common::Pipeline<Communicator>::ReadOnlySender;
+using CommunicatorPipelineReadWriteSender = typename common::Pipeline<Communicator>::ReadWriteSender;
+
+/// A CommunicatorPipeline provides pipelined access to a communicator, as well as metadata about the
+/// communicator.
+///
+/// A CommunicatorPipeline uses a Pipeline to provide shared or exclusive access through senders to the
+/// underlying communicator.  @tparam CT signals whether the contained communicator is a row, column, or
+/// full grid communicator.
+template <CommunicatorType CT>
+class CommunicatorPipeline {
+  using PipelineType = dlaf::common::Pipeline<Communicator>;
+
+public:
+  using ReadOnlyWrapper = typename PipelineType::ReadOnlyWrapper;
+  using ReadWriteWrapper = typename PipelineType::ReadWriteWrapper;
+  using ReadOnlySender = typename PipelineType::ReadOnlySender;
+  using ReadWriteSender = typename PipelineType::ReadWriteSender;
+
+  static constexpr CommunicatorType communicator_type = CT;
+
+  /// Create a CommunicatorPipeline by moving in the resource (it takes the
+  /// ownership).
+  explicit CommunicatorPipeline(Communicator comm, Index2D rank = {0, 0}, Size2D size = {0, 0})
+      : pipeline_(std::move(comm)), rank_(std::move(rank)), size_(std::move(size)) {}
+  CommunicatorPipeline(CommunicatorPipeline&& other) = default;
+  CommunicatorPipeline& operator=(CommunicatorPipeline&& other) = default;
+  CommunicatorPipeline(const CommunicatorPipeline&) = delete;
+  CommunicatorPipeline& operator=(const CommunicatorPipeline&) = delete;
+
+  /// Return the rank of the current process in the CommunicatorPipeline.
+  IndexT_MPI rank() const noexcept {
+    return rankFullCommunicator(rank_);
+  }
+
+  /// Return the size of the grid.
+  IndexT_MPI size() const noexcept {
+    return size_.rows() * size_.cols();
+  }
+
+  /// Return the 2D rank of the current process in the grid that this pipeline
+  /// belongs to.
+  Index2D rank_2d() const noexcept {
+    return rank_;
+  }
+
+  /// Return the 2D size of the current process in the grid that this pipeline
+  /// belongs to.
+  Size2D size_2d() const noexcept {
+    return size_;
+  }
+
+  /// Return rank in the grid with all ranks given the 2D index.
+  IndexT_MPI rankFullCommunicator(const Index2D& index) const noexcept {
+    return common::computeLinearIndex<IndexT_MPI>(internal::FULL_COMMUNICATOR_ORDER, index,
+                                                  {size_.rows(), size_.cols()});
+  }
```

Comment:
I think these would need another pair of eyes. In particular, should e.g. `size` be dependent on the `CommunicatorType`:
```
IndexT_MPI size() const noexcept {
  if constexpr (communicator_type == CommunicatorType::Row) {
    return size_.cols();
  } else if (communicator_type == CommunicatorType::Col) {
    return size_.rows();
  } else {
    return size_.rows() * size_.cols();
  }
}
```
or is even that wrong and I should be getting it from the communicator itself during construction?

Reply:
I think I've answered my own question along the way. I'm now getting the 1d size and rank from the `Communicator` before putting it into a `Pipeline`. Does that sound right?